pkg/objstore/swift/swift.go

```diff
@@ -82,215 +206,141 @@ func (c *Container) Name() string {
 func (c *Container) IsObjNotFoundErr(err error) bool {
-	_, ok := err.(gophercloud.ErrDefault404)
-	return ok
+	return errors.Is(err, swift.ObjectNotFound)
 }
 
 // Upload writes the contents of the reader as an object into the container.
-func (c *Container) Upload(ctx context.Context, name string, r io.Reader) error {
-	options := &objects.CreateOpts{Content: r}
-	res := objects.Create(c.client, c.name, name, options)
-	return res.Err
-}
-
-// Delete removes the object with the given name.
-func (c *Container) Delete(ctx context.Context, name string) error {
-	return objects.Delete(c.client, c.name, name, nil).Err
-}
-
-func (*Container) Close() error {
-	// Nothing to close.
-	return nil
-}
-
-func parseConfig(conf []byte) (*SwiftConfig, error) {
-	var sc SwiftConfig
-	err := yaml.UnmarshalStrict(conf, &sc)
-	return &sc, err
-}
-
-func authOptsFromConfig(sc *SwiftConfig) gophercloud.AuthOptions {
-	authOpts := gophercloud.AuthOptions{
-		IdentityEndpoint: sc.AuthUrl,
-		Username:         sc.Username,
-		UserID:           sc.UserId,
-		Password:         sc.Password,
-		DomainID:         sc.DomainId,
-		DomainName:       sc.DomainName,
-		TenantID:         sc.ProjectID,
-		TenantName:       sc.ProjectName,
-
-		// Allow Gophercloud to re-authenticate automatically.
-		AllowReauth: true,
+func (c *Container) Upload(_ context.Context, name string, r io.Reader) error {
+	size, err := objstore.TryToGetSize(r)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "could not guess file size, using large object to avoid issues if the file is larger than limit", "name", name, "err", err)
+		// Anything higher or equal to chunk size so the SLO is used.
+		size = c.chunkSize
 	}
-
-	// Support for cross-domain scoping (user in different domain than project).
-	// If a userDomainName or userDomainID is given, the user is scoped to this domain.
-	switch {
-	case sc.UserDomainName != "":
-		authOpts.DomainName = sc.UserDomainName
-	case sc.UserDomainID != "":
-		authOpts.DomainID = sc.UserDomainID
+	var file io.WriteCloser
+	if size >= c.chunkSize {
+		file, err = c.connection.StaticLargeObjectCreateFile(&swift.LargeObjectOpts{
+			Container:        c.name,
+			ObjectName:       name,
+			ChunkSize:        c.chunkSize,
+			SegmentContainer: c.segmentsContainer,
+			CheckHash:        true,
+		})
+	} else {
+		file, err = c.connection.ObjectCreate(c.name, name, true, "", "", swift.Headers{})
 	}
-
-	// A token can be scoped to a domain or project.
-	// The project can be in another domain than the user, which is indicated by setting either projectDomainName or projectDomainID.
-	switch {
-	case sc.ProjectDomainName != "":
-		authOpts.Scope = &gophercloud.AuthScope{
-			DomainName: sc.ProjectDomainName,
+	if err != nil {
+		return errors.Wrap(err, "swift failed to create file")
+	}
+	defer runutil.CloseWithLogOnErr(c.logger, file, "swift upload obj close")
+	switch f := file.(type) {
```

Comment:
Is there a reason not to use `io.Copy()`?

Reply:
TBH the PR is quite old and I do not remember but I do not see any obvious reason not to :thinking: 

Changed it to `io.Copy` and will see hot the tests end but definitely good point :)